<div class="alert alert-block alert-info">

# Especialització de DataScience de la IT ACADEMY
### *Sprint4 - Exercici 1*
#### Estandaritza, identifica i enumera cada un dels atributs / variables de l'estructura de l'arxiu "Web_access_log-akumenius.com" que trobaràs al repositori de GitHub "Data-sources". 
https://github.com/IT-Academy-BCN/Data-Science/blob/main/Data-sources/Web_access_log-akumenius.com.txt.7z

#### Alumne: Jorge Prieto Rey
    
<div>

In [43]:
import pandas as pd
import numpy as np
import re

In [44]:
#Llibreria per extreure fitxer .7z
import py7zr

In [45]:
with py7zr.SevenZipFile('Web_access_log-akumenius.com.txt.7z', mode='r') as z:
    z.extractall()

In [46]:
filename='Web_access_log-akumenius.com.txt'

In [47]:
fields = ['c1','c2','c3','c4','c5','c6','c7','c8','c9','c10','c11','c12','c13','c14']
    
data = pd.read_csv(filename, sep=' ', header=None, names=fields, na_values=['-'])
data

P:\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (7,12) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


c1             c2  c3   c4                     c5  \
0               localhost      127.0.0.1 NaN  NaN  [23/Feb/2014:03:10:31   
1               localhost      127.0.0.1 NaN  NaN  [23/Feb/2014:03:10:31   
2               localhost      127.0.0.1 NaN  NaN  [23/Feb/2014:03:10:31   
3               localhost      127.0.0.1 NaN  NaN  [23/Feb/2014:03:10:31   
4               localhost      127.0.0.1 NaN  NaN  [23/Feb/2014:03:10:31   
...                   ...            ...  ..  ...                    ...   
261868  www.akumenius.com   5.255.253.53 NaN  NaN  [02/Mar/2014:03:05:39   
261869  www.akumenius.com  74.86.158.107 NaN  NaN  [02/Mar/2014:03:09:52   
261870          localhost      127.0.0.1 NaN  NaN  [02/Mar/2014:03:10:18   
261871          localhost      127.0.0.1 NaN  NaN  [02/Mar/2014:03:10:18   
261872          localhost      127.0.0.1 NaN  NaN  [02/Mar/2014:03:10:18   

            c6                  c7   c8      c9  c10  \
0       +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   
1       +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   
2       +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   
3       +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   
4       +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   
...        ...                 ...  ...     ...  ...   
261868  +0100]      GET / HTTP/1.1  200  7528.0  NaN   
261869  +0100]     HEAD / HTTP/1.1  200     NaN  NaN   
261870  +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   
261871  +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   
261872  +0100]  OPTIONS * HTTP/1.0  200     NaN  NaN   

                                                      c11     c12  c13  c14  
0                      Apache (internal dummy connection)  VLOG=-  NaN  NaN  
1                      Apache (internal dummy connection)  VLOG=-  NaN  NaN  
2                      Apache (internal dummy connection)  VLOG=-  NaN  NaN  
3                      Apache (internal dummy connection)  VLOG=-  NaN  NaN  
4                      Apache (internal dummy connection)  VLOG=-  NaN  NaN  
...                                                   ...     ...  ...  ...  
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:...  VLOG=-  NaN  NaN  
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http...  VLOG=-  NaN  NaN  
261870                 Apache (internal dummy connection)  VLOG=-  NaN  NaN  
261871                 Apache (internal dummy connection)  VLOG=-  NaN  NaN  
261872                 Apache (internal dummy connection)  VLOG=-  NaN  NaN  

[261873 rows x 14 columns]

<div class="alert alert-block alert-info">

### *Sprint4 - Exercici 2*

Neteja, preprocesa, estructura i transforma (dataframe) les dades del registre d'Accés a la web. 
    
<div>

In [48]:
print('Total files: '+str(len(data)))
print('nombre de NaN en Columnes:')
data.isna().sum()

Total files: 261873
nombre de NaN en Columnes:


c1          0
c2          0
c3     261873
c4     261846
c5          0
c6          0
c7         37
c8          0
c9      42335
c10     99545
c11       221
c12         0
c13    261871
c14    261873
dtype: int64

In [49]:
#eliminem la columna c3 i c14 ja que no conte dades
data=data.drop(['c3','c14'],axis=1)

In [50]:
#renombrem les columnes del dataframe
newLength = len(fields)-2;
fields= fields[0:newLength]
data.columns = fields
data

c1             c2   c3                     c4      c5  \
0               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
1               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
2               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
3               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
4               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
...                   ...            ...  ...                    ...     ...   
261868  www.akumenius.com   5.255.253.53  NaN  [02/Mar/2014:03:05:39  +0100]   
261869  www.akumenius.com  74.86.158.107  NaN  [02/Mar/2014:03:09:52  +0100]   
261870          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261871          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261872          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   

                        c6   c7      c8   c9  \
0       OPTIONS * HTTP/1.0  200     NaN  NaN   
1       OPTIONS * HTTP/1.0  200     NaN  NaN   
2       OPTIONS * HTTP/1.0  200     NaN  NaN   
3       OPTIONS * HTTP/1.0  200     NaN  NaN   
4       OPTIONS * HTTP/1.0  200     NaN  NaN   
...                    ...  ...     ...  ...   
261868      GET / HTTP/1.1  200  7528.0  NaN   
261869     HEAD / HTTP/1.1  200     NaN  NaN   
261870  OPTIONS * HTTP/1.0  200     NaN  NaN   
261871  OPTIONS * HTTP/1.0  200     NaN  NaN   
261872  OPTIONS * HTTP/1.0  200     NaN  NaN   

                                                      c10     c11  c12  
0                      Apache (internal dummy connection)  VLOG=-  NaN  
1                      Apache (internal dummy connection)  VLOG=-  NaN  
2                      Apache (internal dummy connection)  VLOG=-  NaN  
3                      Apache (internal dummy connection)  VLOG=-  NaN  
4                      Apache (internal dummy connection)  VLOG=-  NaN  
...                                                   ...     ...  ...  
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:...  VLOG=-  NaN  
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http...  VLOG=-  NaN  
261870                 Apache (internal dummy connection)  VLOG=-  NaN  
261871                 Apache (internal dummy connection)  VLOG=-  NaN  
261872                 Apache (internal dummy connection)  VLOG=-  NaN  

[261873 rows x 12 columns]

In [51]:
data['c12'].value_counts(sort=False)

VLOG=-    2
Name: c12, dtype: int64

In [52]:
data1 = data.loc[data['c12'] == 'VLOG=-']

In [53]:
data1.index

Int64Index([181473, 236496], dtype='int64')

#### Hi ha 2 files que no estan ben alineades perquè 'VLOG=-' esta a la columna c13

In [54]:
data[data1.index[0]-1:data1.index[0]+2]

c1              c2   c3                     c4      c5  \
181472  www.akumenius.com   74.86.158.106  NaN  [26/Feb/2014:23:44:15  +0100]   
181473  www.akumenius.com      5.10.83.41  NaN  [26/Feb/2014:23:44:22  +0100]   
181474  www.akumenius.com  220.181.108.96  NaN  [26/Feb/2014:23:44:31  +0100]   

                                                       c6         c7       c8  \
181472                                    HEAD / HTTP/1.1        200      NaN   
181473  GET /escapadas/escapada-especial-san-valentin-...  HTTP/1.1"    200.0   
181474                                     GET / HTTP/1.1        200  25030.0   

          c9                                                c10  \
181472   NaN  Mozilla/5.0+(compatible; UptimeRobot/2.0; http...   
181473  3100                                                NaN   
181474   NaN  Mozilla/5.0 (compatible; Baiduspider/2.0; +htt...   

                                                      c11     c12  
181472                                             VLOG=-     NaN  
181473  Mozilla/5.0 (compatible; AhrefsBot/5.0; +http:...  VLOG=-  
181474                                             VLOG=-     NaN

In [55]:
data[data1.index[1]-1:data1.index[1]+2]

c1             c2   c3                     c4      c5  \
236495  www.akumenius.com  89.29.152.238  NaN  [28/Feb/2014:16:07:21  +0100]   
236496  www.akumenius.com     5.10.83.11  NaN  [28/Feb/2014:16:07:20  +0100]   
236497  www.akumenius.com  89.29.152.238  NaN  [28/Feb/2014:16:07:22  +0100]   

                                                       c6         c7     c8  \
236495                     POST /destinos-insert HTTP/1.1        200   26.0   
236496  GET /escapadas/escapada-\al-salto-de-roldan\"-...  HTTP/1.1"  200.0   
236497  GET /libraries/jqueryui/css/ui-lightness/image...        200  104.0   

                               c9  \
236495  http://www.akumenius.com/   
236496                       3100   
236497  http://www.akumenius.com/   

                                                      c10  \
236495  Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.3...   
236496                                                NaN   
236497  Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.3...   

                                                      c11     c12  
236495                                             VLOG=-     NaN  
236496  Mozilla/5.0 (compatible; AhrefsBot/5.0; +http:...  VLOG=-  
236497                                             VLOG=-     NaN

In [56]:
#Juntem les columnes C6 i C7
data1['c6'] = data1['c6'].map(str)+""+data1['c7']

P:\anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [57]:
#Eliminen la columna C7
data1=data1.drop(['c7'],axis=1)
data1

c1          c2   c3                     c4      c5  \
181473  www.akumenius.com  5.10.83.41  NaN  [26/Feb/2014:23:44:22  +0100]   
236496  www.akumenius.com  5.10.83.11  NaN  [28/Feb/2014:16:07:20  +0100]   

                                                       c6     c8    c9  c10  \
181473  GET /escapadas/escapada-especial-san-valentin-...  200.0  3100  NaN   
236496  GET /escapadas/escapada-\al-salto-de-roldan\"-...  200.0  3100  NaN   

                                                      c11     c12  
181473  Mozilla/5.0 (compatible; AhrefsBot/5.0; +http:...  VLOG=-  
236496  Mozilla/5.0 (compatible; AhrefsBot/5.0; +http:...  VLOG=-

In [58]:
data1.loc[:,'b1'] = np.nan
#renombrem les columnes del dataframe
data1.columns = fields
data1

c1          c2   c3                     c4      c5  \
181473  www.akumenius.com  5.10.83.41  NaN  [26/Feb/2014:23:44:22  +0100]   
236496  www.akumenius.com  5.10.83.11  NaN  [28/Feb/2014:16:07:20  +0100]   

                                                       c6     c7    c8   c9  \
181473  GET /escapadas/escapada-especial-san-valentin-...  200.0  3100  NaN   
236496  GET /escapadas/escapada-\al-salto-de-roldan\"-...  200.0  3100  NaN   

                                                      c10     c11  c12  
181473  Mozilla/5.0 (compatible; AhrefsBot/5.0; +http:...  VLOG=-  NaN  
236496  Mozilla/5.0 (compatible; AhrefsBot/5.0; +http:...  VLOG=-  NaN

In [59]:
#reemplaçem les files 181473 i 236496
data.iloc[data1.index]=data1
data

c1             c2   c3                     c4      c5  \
0               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
1               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
2               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
3               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
4               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
...                   ...            ...  ...                    ...     ...   
261868  www.akumenius.com   5.255.253.53  NaN  [02/Mar/2014:03:05:39  +0100]   
261869  www.akumenius.com  74.86.158.107  NaN  [02/Mar/2014:03:09:52  +0100]   
261870          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261871          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261872          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   

                        c6   c7    c8   c9  \
0       OPTIONS * HTTP/1.0  200   NaN  NaN   
1       OPTIONS * HTTP/1.0  200   NaN  NaN   
2       OPTIONS * HTTP/1.0  200   NaN  NaN   
3       OPTIONS * HTTP/1.0  200   NaN  NaN   
4       OPTIONS * HTTP/1.0  200   NaN  NaN   
...                    ...  ...   ...  ...   
261868      GET / HTTP/1.1  200  7528  NaN   
261869     HEAD / HTTP/1.1  200   NaN  NaN   
261870  OPTIONS * HTTP/1.0  200   NaN  NaN   
261871  OPTIONS * HTTP/1.0  200   NaN  NaN   
261872  OPTIONS * HTTP/1.0  200   NaN  NaN   

                                                      c10     c11  c12  
0                      Apache (internal dummy connection)  VLOG=-  NaN  
1                      Apache (internal dummy connection)  VLOG=-  NaN  
2                      Apache (internal dummy connection)  VLOG=-  NaN  
3                      Apache (internal dummy connection)  VLOG=-  NaN  
4                      Apache (internal dummy connection)  VLOG=-  NaN  
...                                                   ...     ...  ...  
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:...  VLOG=-  NaN  
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http...  VLOG=-  NaN  
261870                 Apache (internal dummy connection)  VLOG=-  NaN  
261871                 Apache (internal dummy connection)  VLOG=-  NaN  
261872                 Apache (internal dummy connection)  VLOG=-  NaN  

[261873 rows x 12 columns]

In [60]:
#Verifiquem que la columna c12 no te valors
if ((len(data['c12'])) == (data['c12'].isnull().sum(axis=0))):
    print('Columna buida')

Columna buida


In [61]:
#eliminem la columna c12 ja que no conte dades
data=data.drop(['c12'],axis=1)
data

c1             c2   c3                     c4      c5  \
0               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
1               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
2               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
3               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
4               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
...                   ...            ...  ...                    ...     ...   
261868  www.akumenius.com   5.255.253.53  NaN  [02/Mar/2014:03:05:39  +0100]   
261869  www.akumenius.com  74.86.158.107  NaN  [02/Mar/2014:03:09:52  +0100]   
261870          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261871          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261872          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   

                        c6   c7    c8   c9  \
0       OPTIONS * HTTP/1.0  200   NaN  NaN   
1       OPTIONS * HTTP/1.0  200   NaN  NaN   
2       OPTIONS * HTTP/1.0  200   NaN  NaN   
3       OPTIONS * HTTP/1.0  200   NaN  NaN   
4       OPTIONS * HTTP/1.0  200   NaN  NaN   
...                    ...  ...   ...  ...   
261868      GET / HTTP/1.1  200  7528  NaN   
261869     HEAD / HTTP/1.1  200   NaN  NaN   
261870  OPTIONS * HTTP/1.0  200   NaN  NaN   
261871  OPTIONS * HTTP/1.0  200   NaN  NaN   
261872  OPTIONS * HTTP/1.0  200   NaN  NaN   

                                                      c10     c11  
0                      Apache (internal dummy connection)  VLOG=-  
1                      Apache (internal dummy connection)  VLOG=-  
2                      Apache (internal dummy connection)  VLOG=-  
3                      Apache (internal dummy connection)  VLOG=-  
4                      Apache (internal dummy connection)  VLOG=-  
...                                                   ...     ...  
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:...  VLOG=-  
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http...  VLOG=-  
261870                 Apache (internal dummy connection)  VLOG=-  
261871                 Apache (internal dummy connection)  VLOG=-  
261872                 Apache (internal dummy connection)  VLOG=-  

[261873 rows x 11 columns]

In [62]:
data['c11'].value_counts(sort=False)

VLOG=-    261873
Name: c11, dtype: int64

In [63]:
#eliminem la columna c11 ja que totes les columnes contenen el mateix valors
data=data.drop(['c11'],axis=1)

#### Renombren les columnes amb la descripció dels elements del log ####

In [64]:
data

c1             c2   c3                     c4      c5  \
0               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
1               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
2               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
3               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
4               localhost      127.0.0.1  NaN  [23/Feb/2014:03:10:31  +0100]   
...                   ...            ...  ...                    ...     ...   
261868  www.akumenius.com   5.255.253.53  NaN  [02/Mar/2014:03:05:39  +0100]   
261869  www.akumenius.com  74.86.158.107  NaN  [02/Mar/2014:03:09:52  +0100]   
261870          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261871          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   
261872          localhost      127.0.0.1  NaN  [02/Mar/2014:03:10:18  +0100]   

                        c6   c7    c8   c9  \
0       OPTIONS * HTTP/1.0  200   NaN  NaN   
1       OPTIONS * HTTP/1.0  200   NaN  NaN   
2       OPTIONS * HTTP/1.0  200   NaN  NaN   
3       OPTIONS * HTTP/1.0  200   NaN  NaN   
4       OPTIONS * HTTP/1.0  200   NaN  NaN   
...                    ...  ...   ...  ...   
261868      GET / HTTP/1.1  200  7528  NaN   
261869     HEAD / HTTP/1.1  200   NaN  NaN   
261870  OPTIONS * HTTP/1.0  200   NaN  NaN   
261871  OPTIONS * HTTP/1.0  200   NaN  NaN   
261872  OPTIONS * HTTP/1.0  200   NaN  NaN   

                                                      c10  
0                      Apache (internal dummy connection)  
1                      Apache (internal dummy connection)  
2                      Apache (internal dummy connection)  
3                      Apache (internal dummy connection)  
4                      Apache (internal dummy connection)  
...                                                   ...  
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:...  
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http...  
261870                 Apache (internal dummy connection)  
261871                 Apache (internal dummy connection)  
261872                 Apache (internal dummy connection)  

[261873 rows x 10 columns]

In [65]:
data['c3'].value_counts(sort=False)

clarcat    27
Name: c3, dtype: int64

In [66]:
data1 = data.loc[data['c3'] == 'clarcat']
data1

c1             c2       c3                     c4  \
51797   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:54:07   
51799   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:54:08   
51809   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:54:08   
51905   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:55:34   
51906   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:55:34   
51907   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:55:34   
51923   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:55:54   
51924   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:55:54   
51926   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:55:54   
51939   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:56:13   
51940   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:56:13   
51943   www.akumenius.com  80.28.221.123  clarcat  [24/Feb/2014:09:56:14   
97210   www.akumenius.com  80.28.221.123  clarcat  [25/Feb/2014:09:53:27   
97211   www.akumenius.com  80.28.221.123  clarcat  [25/Feb/2014:09:53:28   
97215   www.akumenius.com  80.28.221.123  clarcat  [25/Feb/2014:09:53:28   
153621  www.akumenius.com  80.28.221.123  clarcat  [26/Feb/2014:10:02:46   
153622  www.akumenius.com  80.28.221.123  clarcat  [26/Feb/2014:10:02:46   
153631  www.akumenius.com  80.28.221.123  clarcat  [26/Feb/2014:10:02:46   
195142  www.akumenius.com  80.28.221.123  clarcat  [27/Feb/2014:12:48:52   
195143  www.akumenius.com  80.28.221.123  clarcat  [27/Feb/2014:12:48:52   
195149  www.akumenius.com  80.28.221.123  clarcat  [27/Feb/2014:12:48:52   
196484  www.akumenius.com  80.28.221.123  clarcat  [27/Feb/2014:13:43:05   
196486  www.akumenius.com  80.28.221.123  clarcat  [27/Feb/2014:13:43:06   
196488  www.akumenius.com  80.28.221.123  clarcat  [27/Feb/2014:13:43:06   
230605  www.akumenius.com  80.28.221.123  clarcat  [28/Feb/2014:09:41:38   
230606  www.akumenius.com  80.28.221.123  clarcat  [28/Feb/2014:09:41:39   
230610  www.akumenius.com  80.28.221.123  clarcat  [28/Feb/2014:09:41:39   

            c5                                                 c6   c7  \
51797   +0100]                              GET /awstats HTTP/1.1  200   
51799   +0100]        GET /awstats.pl?framename=mainleft HTTP/1.1  200   
51809   +0100]       GET /awstats.pl?framename=mainright HTTP/1.1  200   
51905   +0100]  GET /awstats.pl?month=02&year=2013&output=main...  200   
51906   +0100]  GET /awstats.pl?month=02&year=2013&output=main...  200   
51907   +0100]  GET /awstats.pl?month=02&year=2013&output=main...  200   
51923   +0100]  GET /awstats.pl?month=02&year=2012&output=main...  200   
51924   +0100]  GET /awstats.pl?month=02&year=2012&output=main...  200   
51926   +0100]  GET /awstats.pl?month=02&year=2012&output=main...  200   
51939   +0100]  GET /awstats.pl?month=02&year=2014&output=main...  200   
51940   +0100]  GET /awstats.pl?month=02&year=2014&output=main...  200   
51943   +0100]  GET /awstats.pl?month=02&year=2014&output=main...  200   
97210   +0100]                              GET /awstats HTTP/1.1  200   
97211   +0100]        GET /awstats.pl?framename=mainleft HTTP/1.1  200   
97215   +0100]       GET /awstats.pl?framename=mainright HTTP/1.1  200   
153621  +0100]                              GET /awstats HTTP/1.1  200   
153622  +0100]        GET /awstats.pl?framename=mainleft HTTP/1.1  200   
153631  +0100]       GET /awstats.pl?framename=mainright HTTP/1.1  200   
195142  +0100]                              GET /awstats HTTP/1.1  200   
195143  +0100]        GET /awstats.pl?framename=mainleft HTTP/1.1  200   
195149  +0100]       GET /awstats.pl?framename=mainright HTTP/1.1  200   
196484  +0100]                              GET /awstats HTTP/1.1  200   
196486  +0100]        GET /awstats.pl?framename=mainleft HTTP/1.1  200   
196488  +0100]       GET /awstats.pl?framename=mainright HTTP/1.1  200   
230605  +0100]                              

In [67]:
#Eliminem la columna c3 perque no ens dona informació rellevant
data=data.drop(['c3'],axis=1)
data

c1             c2                     c4      c5  \
0               localhost      127.0.0.1  [23/Feb/2014:03:10:31  +0100]   
1               localhost      127.0.0.1  [23/Feb/2014:03:10:31  +0100]   
2               localhost      127.0.0.1  [23/Feb/2014:03:10:31  +0100]   
3               localhost      127.0.0.1  [23/Feb/2014:03:10:31  +0100]   
4               localhost      127.0.0.1  [23/Feb/2014:03:10:31  +0100]   
...                   ...            ...                    ...     ...   
261868  www.akumenius.com   5.255.253.53  [02/Mar/2014:03:05:39  +0100]   
261869  www.akumenius.com  74.86.158.107  [02/Mar/2014:03:09:52  +0100]   
261870          localhost      127.0.0.1  [02/Mar/2014:03:10:18  +0100]   
261871          localhost      127.0.0.1  [02/Mar/2014:03:10:18  +0100]   
261872          localhost      127.0.0.1  [02/Mar/2014:03:10:18  +0100]   

                        c6   c7    c8   c9  \
0       OPTIONS * HTTP/1.0  200   NaN  NaN   
1       OPTIONS * HTTP/1.0  200   NaN  NaN   
2       OPTIONS * HTTP/1.0  200   NaN  NaN   
3       OPTIONS * HTTP/1.0  200   NaN  NaN   
4       OPTIONS * HTTP/1.0  200   NaN  NaN   
...                    ...  ...   ...  ...   
261868      GET / HTTP/1.1  200  7528  NaN   
261869     HEAD / HTTP/1.1  200   NaN  NaN   
261870  OPTIONS * HTTP/1.0  200   NaN  NaN   
261871  OPTIONS * HTTP/1.0  200   NaN  NaN   
261872  OPTIONS * HTTP/1.0  200   NaN  NaN   

                                                      c10  
0                      Apache (internal dummy connection)  
1                      Apache (internal dummy connection)  
2                      Apache (internal dummy connection)  
3                      Apache (internal dummy connection)  
4                      Apache (internal dummy connection)  
...                                                   ...  
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:...  
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http...  
261870                 Apache (internal dummy connection)  
261871                 Apache (internal dummy connection)  
261872                 Apache (internal dummy connection)  

[261873 rows x 9 columns]

#### La data en les columnes c4 i c5 representada el temps en dues part, separem la data i l'hora en una nova columna anomenada time i eliminem la columna de zona horaria. També eliminem les columnes c4 i c5 ####

In [68]:
time = data['c4'] + data['c5']
time_trimmed = time.map(lambda s: s.strip('[]').split('+')[0])
data['time'] = pd.to_datetime(time_trimmed, format='%d/%b/%Y:%H:%M:%S')
data=data.drop(['c4'],axis=1)
data=data.drop(['c5'],axis=1)

#### Separem la columna c6 en 3 columnes per "Protocol" + "command" i "path" ####

In [69]:
data_c6=data['c6'].str.split(expand=True)
data_c6.columns = ['command','path','protocol']
data_c6

command path  protocol
0       OPTIONS    *  HTTP/1.0
1       OPTIONS    *  HTTP/1.0
2       OPTIONS    *  HTTP/1.0
3       OPTIONS    *  HTTP/1.0
4       OPTIONS    *  HTTP/1.0
...         ...  ...       ...
261868      GET    /  HTTP/1.1
261869     HEAD    /  HTTP/1.1
261870  OPTIONS    *  HTTP/1.0
261871  OPTIONS    *  HTTP/1.0
261872  OPTIONS    *  HTTP/1.0

[261873 rows x 3 columns]

In [70]:
#verifiquem que la columna c9 conte les url
data['c9'].value_counts(sort=False)

http://www.google.es/aclk?sa=L&ai=ChnJsZfMOU9T3KaWljAb05oHwAY2dy7gEjba3hVubtKiHCggAEAMoA1CX286vBWDV9diCsCvIAQGpAkO2n5CF9rY-qgQoT9CHU_sHJXX9--ed1eSclahGbOC2xSUjGhymdk_i93K-5NT3YkTtlYAHkcqaE5AHAg&sig=AOD64_1Paps_NYJ1OSqwIEJZf0tycDfFVg&rct=j&q=todo+incluido+mallorca&ved=0CEYQ0Qw&adurl=http://www.akumenius.com/hoteles-baratos/hoteles-todo-incluido.html     1
http://www.akumenius.com/destinos-baratos/hoteles-baratos/ofertas-hotel-Apollonium-Club-La-Costa-Beach-&-Resort-en-Altinkum_TURQUIA-146677b-destinos.html                                                                                                                                                                                                         20
http://www.akumenius.com/hoteles-baratos/ofertas-vacaciones-verano-en-costa-maresme.html                                                                                                                                                                                                      

In [71]:
data=data.drop(['c6'],axis=1) #Eliminar columna c6
data=pd.concat([data,data_c6],axis = 1)

In [72]:
#c7 http_code 
#c8 response bytes
#c10 user_agent
#c9 url
data.rename(columns = {'c1':'host','c2':'ip','c7':'http_code','c8' : 'response_bytes','c9':'url','c10':'user_agent'},inplace = True)
data

host             ip http_code response_bytes  url  \
0               localhost      127.0.0.1       200            NaN  NaN   
1               localhost      127.0.0.1       200            NaN  NaN   
2               localhost      127.0.0.1       200            NaN  NaN   
3               localhost      127.0.0.1       200            NaN  NaN   
4               localhost      127.0.0.1       200            NaN  NaN   
...                   ...            ...       ...            ...  ...   
261868  www.akumenius.com   5.255.253.53       200           7528  NaN   
261869  www.akumenius.com  74.86.158.107       200            NaN  NaN   
261870          localhost      127.0.0.1       200            NaN  NaN   
261871          localhost      127.0.0.1       200            NaN  NaN   
261872          localhost      127.0.0.1       200            NaN  NaN   

                                               user_agent                time  \
0                      Apache (internal dummy connection) 2014-02-23 03:10:31   
1                      Apache (internal dummy connection) 2014-02-23 03:10:31   
2                      Apache (internal dummy connection) 2014-02-23 03:10:31   
3                      Apache (internal dummy connection) 2014-02-23 03:10:31   
4                      Apache (internal dummy connection) 2014-02-23 03:10:31   
...                                                   ...                 ...   
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:... 2014-03-02 03:05:39   
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http... 2014-03-02 03:09:52   
261870                 Apache (internal dummy connection) 2014-03-02 03:10:18   
261871                 Apache (internal dummy connection) 2014-03-02 03:10:18   
261872                 Apache (internal dummy connection) 2014-03-02 03:10:18   

        command path  protocol  
0       OPTIONS    *  HTTP/1.0  
1       OPTIONS    *  HTTP/1.0  
2       OPTIONS    *  HTTP/1.0  
3       OPTIONS    *  HTTP/1.0  
4       OPTIONS    *  HTTP/1.0  
...         ...  ...       ...  
261868      GET    /  HTTP/1.1  
261869     HEAD    /  HTTP/1.1  
261870  OPTIONS    *  HTTP/1.0  
261871  OPTIONS    *  HTTP/1.0  
261872  OPTIONS    *  HTTP/1.0  

[261873 rows x 10 columns]

In [73]:
#exemple de fila del dataframe
data.iloc[15452]

host                                              www.akumenius.com
ip                                                    88.31.143.254
http_code                                                       200
response_bytes                                                  742
url                             http://www.akumenius.com/hotel-list
user_agent        Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKi...
time                                            2014-02-23 12:51:43
command                                                         GET
path              /modules/raton/views/themes/bcoos/images/busca...
protocol                                                   HTTP/1.1
Name: 15452, dtype: object

<div class="alert alert-block alert-info">

### *Sprint4 - Exercici 3*

Geolocalitza les IP's.
    
<div>

In [74]:
import requests
import json

In [75]:
# URL de la API
api_url = "http://ip-api.com/json/"

In [76]:
#lista ips
data_ips=(data['ip'].value_counts(sort=False)).index

In [77]:
data_ips

Index(['173.252.101.119', '180.76.6.29', '157.56.93.187', '217.12.16.56',
       '84.125.143.42', '90.174.1.32', '88.0.211.168', '184.73.59.145',
       '83.55.163.245', '148.251.69.136',
       ...
       '80.58.250.78', '180.76.5.97', '180.76.6.52', '77.226.72.99',
       '157.55.33.26', '81.45.52.199', '81.45.53.212', '84.77.142.252',
       '185.3.117.155', '37.15.118.53'],
      dtype='object', length=2921)

In [78]:
data_ips[0]

'173.252.101.119'

In [79]:
# Definimos los parametros de respuesta que queremos obtener
parametros = 'status,country,countryCode,region,regionName,city,zip,lat,lon,timezone,isp,org,as,query'
data_Ip = {"fields":parametros}

In [80]:
def ip_scraping(ip=""):
 # Nos conectamos con la API
 res = requests.get(api_url+ip, data=data_Ip)
 # Obtenemos y procesamos la respuesta JSON
 api_json_res = json.loads(res.content)
 return api_json_res

In [81]:
ip_response = ip_scraping(data_ips[0])
ip_response

{'status': 'success',
 'country': 'Canada',
 'countryCode': 'CA',
 'region': 'ON',
 'regionName': 'Ontario',
 'city': 'Toronto',
 'zip': 'M5A',
 'lat': 43.6532,
 'lon': -79.3832,
 'timezone': 'America/Toronto',
 'isp': 'Facebook, Inc.',
 'org': 'Facebook, Inc.',
 'as': 'AS32934 Facebook, Inc.',
 'query': '173.252.101.119'}

In [94]:
ip_df=pd.DataFrame([ip_scraping(data_ips[0])])

In [95]:
ip_df

status country countryCode region regionName     city  zip      lat  \
0  success  Canada          CA     ON    Ontario  Toronto  M5A  43.6532   

       lon         timezone             isp             org  \
0 -79.3832  America/Toronto  Facebook, Inc.  Facebook, Inc.   

                       as            query  
0  AS32934 Facebook, Inc.  173.252.101.119

In [96]:
data_ips1 = data_ips[:50]

In [97]:
for ip in data_ips1:
    ip_df = ip_df.append([ip_scraping(ip)],ignore_index=True)

In [98]:
ip_df

status        country countryCode region                  regionName  \
0   success         Canada          CA     ON                     Ontario   
1   success         Canada          CA     ON                     Ontario   
2   success          China          CN     BJ                     Beijing   
3   success  United States          US     WA                  Washington   
4   success          Spain          ES     AN                   Andalusia   
5   success          Spain          ES     AN                   Andalusia   
6   success          Spain          ES     MD                      Madrid   
7   success          Spain          ES     CM          Castille-La Mancha   
8   success  United States          US     VA                    Virginia   
9   success          Spain          ES     MD                      Madrid   
10  success        Germany          DE     SN                      Saxony   
11  success          China          CN     BJ                     Beijing   
12  success          Spain          ES     CM          Castille-La Mancha   
13  success  United States          US     WA                  Washington   
14  success          Spain          ES     CL           Castille and León   
15  success  United States          US     WA                  Washington   
16  success          Spain          ES     MD                      Madrid   
17  success        Bolivia          BO      C  Departamento de Cochabamba   
18  success  United States          US     FL                     Florida   
19  success         Brazil          BR     GO                       Goias   
20  success          Spain          ES     MD                      Madrid   
21  success          Spain          ES     MD                      Madrid   
22  success          Spain          ES     IB            Balearic Islands   
23  success         Canada          CA     ON                     Ontario   
24  success  United States          US     IA                        Iowa   
25  success          Spain          ES     GA                     Galicia   
26  success    Netherlands          NL     NH               North Holland   
27  success        Andorra          AD     07            Andorra la Vella   
28  success          Spain          ES     MD                      Madrid   
29  success          Spain          ES     MD                      Madrid   
30  success         Canada          CA     ON                     Ontario   
31  success         Canada          CA     ON                     Ontario   
32  success        Germany          DE     HE                       Hesse   
33  success  United States          US     WA                  Washington   
34  success          Spain          ES     CN              Canary Islands   
35  success         Canada          CA     ON                     Ontario   
36  success         Mexico          MX    JAL                     Jalisco   
37  success  United States          US     CA                  California   
38  success          Spain          ES     AN                   Andalusia   
39  success         France          FR    IDF               Île-de-France   
40  success         Canada          CA     ON                     Ontario   
41  success          Spain          ES     CL           Castille and León   
42  success          Egypt          EG     GZ                        Giza   
43  success          Spain          ES     GA                     Galicia   
44  success          Spain          ES     GA                     Galicia   
45  success          China          CN     BJ                     Beijing   
46  success          Spain          ES     CN              Canary Islands   
47  success          Spain          ES     PV              Basque Country   
48  success         Mexico          MX    CMX                 Mexico City   
49  success          Spain          ES     MD                      Madrid   
50  success          Spain          ES     MD                      Madrid   



#### La funció no permet processar totes les 2921 Ips

### Relacions amb el dataframe principal

In [99]:
data

host             ip http_code response_bytes  url  \
0               localhost      127.0.0.1       200            NaN  NaN   
1               localhost      127.0.0.1       200            NaN  NaN   
2               localhost      127.0.0.1       200            NaN  NaN   
3               localhost      127.0.0.1       200            NaN  NaN   
4               localhost      127.0.0.1       200            NaN  NaN   
...                   ...            ...       ...            ...  ...   
261868  www.akumenius.com   5.255.253.53       200           7528  NaN   
261869  www.akumenius.com  74.86.158.107       200            NaN  NaN   
261870          localhost      127.0.0.1       200            NaN  NaN   
261871          localhost      127.0.0.1       200            NaN  NaN   
261872          localhost      127.0.0.1       200            NaN  NaN   

                                               user_agent                time  \
0                      Apache (internal dummy connection) 2014-02-23 03:10:31   
1                      Apache (internal dummy connection) 2014-02-23 03:10:31   
2                      Apache (internal dummy connection) 2014-02-23 03:10:31   
3                      Apache (internal dummy connection) 2014-02-23 03:10:31   
4                      Apache (internal dummy connection) 2014-02-23 03:10:31   
...                                                   ...                 ...   
261868  Mozilla/5.0 (compatible; YandexBot/3.0; +http:... 2014-03-02 03:05:39   
261869  Mozilla/5.0+(compatible; UptimeRobot/2.0; http... 2014-03-02 03:09:52   
261870                 Apache (internal dummy connection) 2014-03-02 03:10:18   
261871                 Apache (internal dummy connection) 2014-03-02 03:10:18   
261872                 Apache (internal dummy connection) 2014-03-02 03:10:18   

        command path  protocol  
0       OPTIONS    *  HTTP/1.0  
1       OPTIONS    *  HTTP/1.0  
2       OPTIONS    *  HTTP/1.0  
3       OPTIONS    *  HTTP/1.0  
4       OPTIONS    *  HTTP/1.0  
...         ...  ...       ...  
261868      GET    /  HTTP/1.1  
261869     HEAD    /  HTTP/1.1  
261870  OPTIONS    *  HTTP/1.0  
261871  OPTIONS    *  HTTP/1.0  
261872  OPTIONS    *  HTTP/1.0  

[261873 rows x 10 columns]

In [100]:
http_code_host = data.groupby([data.http_code, data.ip]).count()
http_code_host.iloc[:len(http_code_host.index)]

host  response_bytes  url  user_agent  time  \
http_code ip                                                             
200       107.21.14.116       2               2    0           1     2   
          109.107.105.181    70              70   69          70    70   
          109.120.157.179     2               2    2           2     2   
          109.205.137.141     1               1    0           1     1   
          109.208.120.233     1               1    0           1     1   
...                         ...             ...  ...         ...   ...   
408       85.57.252.60        5               0    0           0     5   
          92.58.211.126       5               0    0           0     5   
500       80.28.221.123       2               0    0           2     2   
502       176.31.255.177     18              18    0          18    18   
          80.28.221.123       1               1    0           1     1   

                           command  path  protocol  
http_code ip                                        
200       107.21.14.116          2     2         2  
          109.107.105.181       70    70        70  
          109.120.157.179        2     2         2  
          109.205.137.141        1     1         1  
          109.208.120.233        1     1         1  
...                            ...   ...       ...  
408       85.57.252.60           0     0         0  
          92.58.211.126          0     0         0  
500       80.28.221.123          2     2         2  
502       176.31.255.177        18    18        18  
          80.28.221.123          1     1         1  

[4564 rows x 8 columns]

### Nombre de peticions per Ip

In [101]:
ip = data.groupby([data.ip]).size()

In [102]:
ip

ip
101.199.108.59      32
101.226.167.226      1
107.178.37.48      194
107.178.38.221     176
107.21.14.116        2
                  ... 
95.63.234.126      265
95.63.60.249       154
96.242.31.129       52
98.137.206.41       11
98.244.10.248       91
Length: 2921, dtype: int64